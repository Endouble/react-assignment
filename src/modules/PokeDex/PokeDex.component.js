import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Card from '../Card';

import { colors, sizes } from '../../styles/vars';

const PokeDex = ({ cards, clickHandler }) => (
    <StyledPokeDex>
        <ul className="pokeDex__list">
            {cards.length ? cards.map(c =>
                <li key={c.id}><Card clickHandler={clickHandler} {...c} /></li>)
                : <li className="pokeDex__list--no">No results</li>
            }
        </ul>
    </StyledPokeDex>
);

/*
    PokeDex Styles
*/
const StyledPokeDex = styled.div`
    background: ${colors.magenta};
    padding: 0 30px 50px;

    .pokeDex__list {
        min-height: calc( 100vh - 240px );
        padding: 0 20px;
        display: flex;
        margin: 0;
        flex-wrap: wrap;
        background: ${colors.blue};
        border-left: 10px solid ${colors.darkPurple};
        border-right: 10px solid ${colors.darkPurple};

        li {
            margin: 0 auto;

            @media (min-width: ${sizes.tablet}) {
                width: 25%;
            }
        }
    }

    li.pokeDex__list--no {
        color: white;
        font-size: 25px;
        margin-top: 15px;
    }
`;

/*
    PokeDex propTypes
*/
PokeDex.propTypes = {
    cards: PropTypes.arrayOf(PropTypes.shape()),
    clickHandler: PropTypes.func,
};
PokeDex.defaultProps = {
    cards: [],
    clickHandler: () => null,
};

export default PokeDex;
