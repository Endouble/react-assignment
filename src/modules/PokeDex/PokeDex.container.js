/* global document */

import React from 'react';
import PropTypes from 'prop-types';

import PokeDexContext from './PokeDex.context';
import Modal from '../Modal';
import CardBio from '../Card/CardBio.component';

class PokeDexData extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            currentCard: {},
            isModalOpen: false,
        };
        this.setCurrentCard = this.setCurrentCard.bind(this);
        this.hideModal = this.hideModal.bind(this);
    }

    componentDidUpdate(prevProps, prevState) {
        const { isModalOpen } = this.state;
        if (isModalOpen !== prevState.isModalOpen) {
            const body = document.querySelector('body');
            body.classList.toggle('body--frozen');
        }
    }

    setCurrentCard(currentCard) {
        this.setState({
            currentCard,
            isModalOpen: true,
        });
    }

    hideModal() {
        this.setState({ isModalOpen: false });
    }

    render() {
        const { isModalOpen, currentCard } = this.state;
        return (
            <div>
                <PokeDexContext
                    setCurrentCard={this.setCurrentCard}
                    setCards={this.props.setCards}
                    toggleIsLoading={this.props.toggleIsLoading}
                />
                <Modal isOpen={isModalOpen} hideModal={this.hideModal}>
                    <CardBio {...currentCard} />
                </Modal>
            </div>
        );
    }
}

/*
    PokeDexData propTypes
*/
PokeDexData.propTypes = {
    setCards: PropTypes.func,
    toggleIsLoading: PropTypes.func,
};

PokeDexData.defaultProps = {
    setCards: () => null,
    toggleIsLoading: () => null,
};

export default PokeDexData;
