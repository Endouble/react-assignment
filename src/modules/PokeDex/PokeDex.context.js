import React from 'react';
import PropTypes from 'prop-types';

import PokeDex from './PokeDex.component';
import { CardsContext, FilterContext } from '../../context';

import getPokemons from '../../api/pokemon';

class PokeDexContext extends React.Component {
    constructor(props) {
        super(props);
        this.filteredCards = this.filteredCards.bind(this);
    }

    componentDidMount() {
        const { toggleIsLoading, setCards } = this.props;
        toggleIsLoading(true);
        getPokemons()
            .then((data) => {
                if (data && data.cards) {
                    setCards(data.cards);
                    toggleIsLoading(false);
                }
            });
    }

    filteredCards(filter, cards) {
        const updatedCards = filter.value !== '' ? cards
            .filter(card =>
                card[filter.filterBy].toLowerCase()
                    .includes(filter.value.toLowerCase())) :
            cards;
        return <PokeDex clickHandler={this.props.setCurrentCard} cards={updatedCards} />;
    }

    render() {
        return (
            <FilterContext.Consumer>
                {filter => (
                    <CardsContext.Consumer>
                        {cards => this.filteredCards(filter, cards)}
                    </CardsContext.Consumer>
                )}
            </FilterContext.Consumer>
        );
    }
}

/*
    PokeDexContext propTypes
*/
PokeDexContext.propTypes = {
    setCards: PropTypes.func,
    setCurrentCard: PropTypes.func,
    toggleIsLoading: PropTypes.func,
};

PokeDexContext.defaultProps = {
    setCards: () => null,
    setCurrentCard: () => null,
    toggleIsLoading: () => null,
};

export default PokeDexContext;
