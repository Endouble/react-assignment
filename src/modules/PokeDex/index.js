import PokeDex from './PokeDex.container';

export default PokeDex;
