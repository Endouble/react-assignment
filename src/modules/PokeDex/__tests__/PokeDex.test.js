import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import PokeDex from '../';

configure({ adapter: new Adapter() });

describe('PokeDex', () => {
    it('renders correctly', () => {
        const component = shallow(<PokeDex />);
        expect(component).toMatchSnapshot();
    });
});
