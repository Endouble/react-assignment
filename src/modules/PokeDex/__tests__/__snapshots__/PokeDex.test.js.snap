// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeDex renders correctly 1`] = `
<div>
  <PokeDexContext
    setCards={[Function]}
    setCurrentCard={[Function]}
    toggleIsLoading={[Function]}
  />
  <Modal
    hideModal={[Function]}
    isOpen={false}
  />
</div>
`;
